lib/CleanPlugin.js

```diff
@@ -0,0 +1,147 @@
+/** @typedef {import("./Compiler")} Compiler */
+
+/**
+ * @param {import("./util/fs").OutputFileSystem} outputFS output fs
+ * @param {string} from from directory
+ * @returns {{files: Array<string>, directories: Array<string>}} collected files and directories
+ */
+function readDir(outputFS, from) {
+	const collectedFiles = [];
+	const collectedDirectories = [];
+	const stack = [from];
+	let cursor;
+
+	while ((cursor = stack.pop())) {
+		const stat = outputFS.statSync(cursor);
+
+		if (stat.isDirectory()) {
+			const items = outputFS.readdirSync(cursor);
+
+			if (from !== cursor) {
+				const relative = path.relative(from, cursor);
+				collectedDirectories.push(relative);
+			}
+
+			for (let i = 0; i < items.length; i++) {
+				stack.push(path.join(cursor, items[i]));
+			}
+		} else {
+			const relative = path.relative(from, cursor);
+			collectedFiles.push(relative);
+		}
+	}
+
+	return {
+		files: collectedFiles,
+		directories: collectedDirectories
+	};
+}
+
+class CleanPlugin {
+	/**
+	 * @param {CleanPluginArgument} [options] options
+	 */
+	constructor(options) {
+		this.options = Object.assign({ enabled: true, dry: false }, options || {});
+
+		if (options && typeof options === "object") {
+			validateOptions(schema, options, {
+				name: "Clean Plugin",
+				baseDataPath: "options"
+			});
+		}
+	}
+
+	/**
+	 * @param {Compiler} compiler the compiler instance
+	 * @returns {void}
+	 */
+	apply(compiler) {
+		let emittedOnce = false;
+		const logger = compiler.getInfrastructureLogger("webpack.Clean");
+		const handleDir = (outputPath, directory) => {
+			if (this.options.dry) {
+				logger.info(`Directory [${directory}] will be removed in non-dry mode`);
+			} else {
+				const abs = path.join(outputPath, directory);
+				if (compiler.outputFileSystem.existsSync(abs)) {
+					compiler.outputFileSystem.rmdirSync(abs, {
+						recursive: true
+					});
+				}
+			}
+		};
+		const handleFile = (outputPath, file) => {
+			if (this.options.dry) {
+				logger.info(`Asset [${file}] will be removed in non-dry mode`);
+			} else {
+				const abs = path.join(outputPath, file);
+				compiler.outputFileSystem.unlinkSync(abs);
+			}
+		};
+
+		// do it before emitting for not to iterate many files
+		compiler.hooks.emit.tap(
+			{
+				name: "CleanPlugin",
+				stage: 100
+			},
+			compilation => {
+				if (emittedOnce) {
+					return;
+				}
+
+				const outputPath = compilation.getPath(compiler.outputPath, {});
+
+				if (
+					!compiler.outputFileSystem ||
+					!compiler.outputFileSystem.existsSync(outputPath)
+				) {
+					return;
+				}
+
+				const assetsDirectories = new Set();
+
+				for (const asset in compilation.assets) {
+					const parts = asset.split(path.sep).slice(0, -1);
+
+					parts.reduce((all, part) => {
+						const directory = path.join(all, part);
+						assetsDirectories.add(directory);
+						return directory;
+					}, "");
+				}
+
+				const collected = readDir(compiler.outputFileSystem, outputPath);
```

Comment:
> It's unnecessary to read the whole fs structure

Then how we know about files and directories that not a webpack-assets? `assetsDirectories` formed from `compilation.assets`

Reply:
> `assetsDirectories` contains directories only from `compilation.assets`

Yep that is perfect. You only need to read these directories to have full information what need to be done.
Read any other directory that's not in assetsDirectories would be wasted as it would be completely deleted anyway as the folder is extraneous in some directory of `assetsDirectories` anyway.